crates/mpz-ole-core/src/receiver.rs

```diff
@@ -0,0 +1,143 @@
+//! Receiver implementation.
+
+use crate::{
+    core::{ReceiverAdjust, ReceiverShare, ShareAdjust},
+    msg::{BatchAdjust, MaskedInputs},
+    OLEError, TransferId,
+};
+use mpz_fields::Field;
+use std::collections::VecDeque;
+
+/// A receiver for batched OLE.
+#[derive(Debug)]
+pub struct OLEReceiver<const N: usize, F> {
+    id: TransferId,
+    cache: VecDeque<ReceiverShare<F>>,
+}
+
+impl<const N: usize, F: Field> Default for OLEReceiver<N, F> {
+    fn default() -> Self {
+        OLEReceiver {
+            id: TransferId::default(),
+            cache: VecDeque::default(),
+        }
+    }
+}
+
+impl<const N: usize, F: Field> OLEReceiver<N, F> {
+    /// Generates new OLEs and stores them internally.
+    ///
+    /// # Arguments
+    ///
+    /// * `input` - The receiver's OLE input factors.
+    /// * `random` - Uniformly random field elements.
+    /// * `masked` - The correlations from the sender.
+    pub fn preprocess(
+        &mut self,
+        input: Vec<F>,
+        random: Vec<F>,
+        masked: MaskedInputs<F>,
+    ) -> Result<(), OLEError> {
+        let masks = masked.try_into()?;
+        let shares = ReceiverShare::new_vec::<N>(input, random, masks)?;
+
+        self.cache.extend(shares);
+        Ok(())
+    }
+
+    /// Returns OLEs from internal cache.
+    ///
+    /// For consumption of OLEs which have been stored by [`OLEReceiver::preprocess`].
+    ///
+    /// # Arguments
+    ///
+    /// * `count` - The number of shares to return.
+    ///
+    /// # Returns
+    ///
+    /// * A vector of [`ReceiverShare`]s containing the OLE outputs for the receiver.
+    pub fn consume(&mut self, count: usize) -> Option<Vec<ReceiverShare<F>>> {
+        if count > self.cache.len() {
+            return None;
+        }
+
+        let shares = self.cache.drain(..count).collect();
+        Some(shares)
+    }
+
+    /// Adjusts OLEs in the internal cache.
+    ///
+    /// # Arguments
+    ///
+    /// * `targets` - The new OLE receiver inputs.
+    ///
+    /// # Returns
+    ///
+    /// * [`BatchReceiverAdjust`] which needs to be converted by [`BatchReceiverAdjust::finish_adjust`].
+    /// * [`BatchAdjust`] which needs to be sent to the [`crate::OLESender`].
```

Comment:
crate::OLESender was called in earlier comments simply "sender"
```suggestion
    /// * [`BatchAdjust`] which needs to be sent to the sender.
```